### PR TITLE
feat: update libp2p-server and metric example to use hyper 1.0 version

### PR DESCRIPTION
**Pull request by getong**
## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by thomaseizinger** [comment]
Sorry for the delay, I should have some time mid-January to review this.

**Comment by getong** [comment]
How about enabling ci first, I'd like to see what ci runs here.

File: examples/metrics/Cargo.toml [hunk 1]
```diff
@@ -10,15 +10,19 @@ release = false
...
 futures = "0.3.30"
-hyper = { version = "0.14", features = ["server", "tcp", "http1"] }
+bytes = "1.5.0"
+hyper = { version = "1.1", features = ["full"]}
+http-body-util = "0.1.0"
+hyper-util = { version = "0.1.2", features = ["full"] }
 libp2p = { path = "../../libp2p", features = ["tokio", "metrics", "ping", "noise", "identify", "tcp", "yamux", "macros"] }
-opentelemetry = { version = "0.20.0", features = ["rt-tokio", "metrics"] }
-opentelemetry-otlp = { version = "0.13.0", features = ["metrics"]}
+opentelemetry = { version = "0.21.0", features = ["metrics"] }
```
Comment (thomaseizinger) [inline examples/metrics/Cargo.toml]: Do we need to update `opentelemetry` in order to update `hyper`? I'd like to keep PRs as focused on a single aspect. If possible, it would be great to do this in a separate PR.

From what I investigated in https://github.com/libp2p/rust-libp2p/actions/runs/7444770125/job/20251786909?pr=4858, there are still some releases missing the `opentelemetry` ecosystem to full make this update.

File: examples/metrics/src/http_service.rs
Comment (thomaseizinger) [inline examples/metrics/src/http_service.rs:1]: It seems that this got even more verbose with the update to `hyper` 1.0. Can we perhaps simplify this by migrating it to a higher-level library like axum?

**Review by thomaseizinger — COMMENTED** [review]
Thank you! I've left some feedback :)

File: examples/metrics/src/http_service.rs
Comment (getong) [reply examples/metrics/src/http_service.rs:1]: Axum is using hyper under the hood, but axum is still not reached 1.0, and hyper now is stable to be used as 1.0.
I think it does not need to change it.

File: examples/metrics/Cargo.toml (hunk 1 above)
Comment (getong) [reply examples/metrics/Cargo.toml]: I update the code by the way,  as opentelemetry api also changes, I think we does not need a seperate pr here.

Comment (thomaseizinger) [reply examples/metrics/Cargo.toml]: > as opentelemetry api also changes

What do you mean by that? Can we update hyper without updating opentelemetry? If yes, then I would like this PR to only deal with updating hyper please.

File: examples/metrics/src/http_service.rs
Comment (thomaseizinger) [reply examples/metrics/src/http_service.rs:1]: We are using many library that are < 1.0. I am more concerned with the boilerplate around setting up an HTTP services with just using `hyper`. As mentioned above, with the update to `1.0`, it seems that that is even more verbose than pre 1.0.

We already use `axum` in other places. Would you be willing to migrate `libp2p-server` and `metric-example` to use latest `axum` (which I believe already uses `hyper` 1.0 under the hood).

**Comment by jxs** [comment]
friendly ping @getong

File: examples/metrics/Cargo.toml (hunk 1 above)
Comment (jxs) [reply examples/metrics/Cargo.toml]: `opentelemetry` has since been updated https://github.com/libp2p/rust-libp2p/pull/5259

File: examples/metrics/src/http_service.rs
Comment (getong) [reply examples/metrics/src/http_service.rs:1]: I have no time to fix it by using axum.

**Comment by getong** [comment]
Quite a long time, with small change updated now.

File: examples/metrics/src/http_service.rs
Comment (jxs) [reply examples/metrics/src/http_service.rs:1]: @drHuangMHT do you wanna give this a try with `axum`?
